src/state_machine/serial.rs

```diff
@@ -68,3 +76,65 @@ where
         Ok(Rc::new(T::decode(target)?))
     }
 }
+
+impl<T: Encode> Encode for Vec<T> {
+    fn encode(&self, target: &mut impl std::io::Write) {
+        target
+            .write_all(&(self.len() as u64).to_le_bytes())
+            .expect("Serialization should not fail");
+        for item in self.iter() {
+            item.encode(target)
+        }
+    }
+}
+
+impl<T: Decode<Error = E>, E> Decode for Vec<T>
+where
+    DeserializationError: From<E>,
+{
+    type Error = DeserializationError;
+
+    fn decode(target: &mut &[u8]) -> Result<Self, Self::Error> {
+        if target.len() < 8 {
+            return Err(DeserializationError::DataTooShort {
+                expected: 8,
+                got: target.len(),
+            });
+        }
+        let mut serialized_len = [0u8; 8];
+        serialized_len.copy_from_slice(&target[..8]);
+        let len: u64 = u64::from_le_bytes(serialized_len);
+        *target = &mut &target[8..];
+
+        let mut out = Vec::new();
+        for _ in 0..len {
+            out.push(T::decode(target)?)
+        }
+        Ok(out)
+    }
+}
+
+impl<T, U> Encode for (T, U)
+where
+    T: Encode,
+    U: Encode,
+{
+    fn encode(&self, target: &mut impl std::io::Write) {
+        self.0.encode(target);
+        self.1.encode(target)
+    }
+}
+
+impl<T, U, E1, E2> Decode for (T, U)
+where
+    T: Decode<Error = E1>,
+    U: Decode<Error = E2>,
+    DeserializationError: From<E1>,
```

Comment:
Here T, U are generic parameters, so we shouldn't make assumptions about their deserialization error type since we might not know it in advance.

What about:

pub enum CombinedError<E1, E2> {
    E1(E1),
    E2(E2),
}

impl<T, U, E1, E2> Decode for (T, U)
where
    T: Decode<Error = E1>,
    U: Decode<Error = E2>,
    CombinedError<E1, E2>: From<E1>,
    CombinedError<E1, E2>: From<E2>,
{
    type Error = CombinedError<E1, E2>;

    fn decode(target: &mut &[u8]) -> Result<Self, Self::Error> {
        Ok((T::decode(target)?, U::decode(target)?))
    }
}

Reply:
Synced with @bkolad offline. We're going to make significant revisions to encode/decode in #18, so I'll remove this piece of the PR